cli/commands/scaffold/action.go

```diff
@@ -0,0 +1,351 @@
+	defer func() {
+		for _, dir := range dirsToClean {
+			if err := os.RemoveAll(dir); err != nil {
+				opts.Logger.Warnf("Failed to clean up dir %s: %v", dir, err)
+			}
+		}
+	}()
+
+	// scaffold only in empty directories
+	if empty, err := util.IsDirectoryEmpty(opts.WorkingDir); !empty || err != nil {
+		if err != nil {
+			return err
+		}
+		return WorkingDirectoryNotEmptyError{
+			dir: opts.WorkingDir,
+		}
+	}
+
+	if len(opts.TerraformCliArgs) >= moduleCount {
+		moduleUrl = opts.TerraformCliArgs[1]
+	}
+
+	if len(opts.TerraformCliArgs) >= moduleAndTemplateCount {
+		templateUrl = opts.TerraformCliArgs[2]
+	}
+
+	tempDir, err := os.MkdirTemp("", "scaffold")
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	dirsToClean = append(dirsToClean, tempDir)
+
+	// prepare variables
+	vars, err := variables.ParseVars(opts.ScaffoldVars, opts.ScaffoldVarFiles)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// parse module url
+	parsedModuleUrl, err := terraform.ToSourceUrl(moduleUrl, tempDir)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	moduleUrl = parsedModuleUrl.String()
+
+	// rewrite module url, if required
+	parsedModuleUrl, err = rewriteModuleUrl(opts, vars, moduleUrl)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// add ref to module url, if required
+	parsedModuleUrl, err = addRefToModuleUrl(opts, parsedModuleUrl, vars)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// regenerate module url with all changes
+	moduleUrl = parsedModuleUrl.String()
+
+	// identify template url
+	templateDir := ""
+	if templateUrl != "" {
+		// process template url if was passed
+		parsedTemplateUrl, err := terraform.ToSourceUrl(templateUrl, tempDir)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		parsedTemplateUrl, err = rewriteTemplateUrl(opts, parsedTemplateUrl)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		// regenerate template url with all changes
+		templateUrl = parsedTemplateUrl.String()
+
+		templateDir, err = os.MkdirTemp("", "template")
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		dirsToClean = append(dirsToClean, templateDir)
+
+		// downloading template
+		opts.Logger.Infof("Using template from %s", templateUrl)
+		if err := getter.GetAny(templateDir, templateUrl); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	opts.Logger.Infof("Scaffolding a new Terragrunt module %s %s to %s", moduleUrl, templateUrl, opts.WorkingDir)
+	if err := getter.GetAny(tempDir, moduleUrl); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// extract variables from module url
+	inputs, err := config.ParseVariables(opts, tempDir)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// separate variables that require value and with default value
+	var requiredVariables []*config.ParsedVariable
+	var optionalVariables []*config.ParsedVariable
+
+	for _, value := range inputs {
+		if value.DefaultValue == "" {
```

Comment:
Does this handle string variables where the default value is an empty string (as opposed to no default value at all)?